packages/ember/tests/service_injection_test.js

```diff
@@ -43,3 +43,131 @@ if (EMBER_METAL_ES5_GETTERS) {
+
+      this.add('controller:route-b', Controller.extend({
+        myService: inject.service('my-service', { source: routeBSource })
+      }));
+
+      let LocalLookupService = Service.extend();
+      this.add({
+        specifier: 'service:my-service',
+        source: routeASource
+      }, LocalLookupService);
+
+      let MyService = Service.extend();
+      this.add({
+        specifier: 'service:my-service',
+        source: routeBSource
+      }, MyService);
+
+      return this.visit('/').then(() => {
+        let controllerA = this.applicationInstance.lookup('controller:route-a');
+        let serviceFromControllerA = controllerA.get('myService');
+        assert.ok(serviceFromControllerA instanceof LocalLookupService, 'local lookup service is returned');
+
+        let controllerB = this.applicationInstance.lookup('controller:route-b');
+        let serviceFromControllerB = controllerB.get('myService');
+        assert.ok(serviceFromControllerB instanceof MyService, 'global service is returned');
+
+        assert.notStrictEqual(serviceFromControllerA, serviceFromControllerB);
+      });
+    }
+
+    ['@test Services can be injected with same name, different source, but same resolution result, and share an instance'](assert) {
+      let routeASource = 'service:src/ui/routes/route-a/controller';
+      let routeBSource = 'service:src/ui/routes/route-b/controller';
+
+      this.add('controller:route-a', Controller.extend({
+        myService: inject.service('my-service', { source: routeASource })
+      }));
+
+      this.add('controller:route-b', Controller.extend({
+        myService: inject.service('my-service', { source: routeBSource })
+      }));
+
+      let MyService = Service.extend();
+      this.add({
+        specifier: 'service:my-service'
+      }, MyService);
+
+      return this.visit('/').then(() => {
+        let controllerA = this.applicationInstance.lookup('controller:route-a');
+        let serviceFromControllerA = controllerA.get('myService');
+        assert.ok(serviceFromControllerA instanceof MyService);
+
+        let controllerB = this.applicationInstance.lookup('controller:route-b');
+        assert.strictEqual(serviceFromControllerA, controllerB.get('myService'));
+      });
+    }
+
+    /*
+     * This test demonstrates a failure in the caching system of ember's
+     * container around singletons and and local lookup. The local lookup
+     * is cached and the global injection is then looked up incorrectly.
+     *
+     * The paractical rules of Ember's module unification config are such
+     * that services cannot be locally looked up, thus this case is really
+     * just a demonstration of what could go wrong if we permit arbitrary
+     * configuration (such as a singleton type that has local lookup).
+     */
+    ['@test Services can be injected with same name, one with source one without, and share an instance'](assert) {
+      let routeASource = 'service:src/ui/routes/route-a/controller';
```

Comment:
Seems odd to have the source here be prefixed with `service:` (the actual type is `controller`)